webapp/api/marketo.py

```diff
@@ -0,0 +1,126 @@
+import os
+
+from webapp import api
+from webapp.api.exceptions import (
+    ApiResponseDecodeError,
+    ApiResponseError,
+    ApiResponseErrorList,
+)
+
+MARKETO_URL = os.getenv("MARKETO_URL", "test.com/")
+MARKETO_CLIENT_ID = os.getenv("MARKETO_CLIENT_ID", "123")
+MARKETO_CLIENT_SECRET = os.getenv("MARKETO_CLIENT_SECRET", "321")
+
+AUTH_URL = "".join(
+    [
+        "https://",
+        MARKETO_URL,
+        "identity/oauth/token?grant_type=client_credentials&client_id=",
+        MARKETO_CLIENT_ID,
+        "&client_secret=",
+        MARKETO_CLIENT_SECRET,
+    ]
+)
+
+LEAD_BY_EMAIL = "".join(
+    [
+        "https://",
+        MARKETO_URL,
+        "rest/v1/leads.json?access_token={token}",
+        "&filterType=email&filterValues={email}&fields=id",
+    ]
+)
+
+LEAD_NEWSLETTER_SUBSCRIPTION = "".join(
+    [
+        "https://",
+        MARKETO_URL,
+        "rest/v1/",
+        "lead/{lead_id}.json?access_token={token}",
+        "&fields=id,email,snapcraftnewsletter",
+    ]
+)
+
+LEADS = "".join(
+    ["https://", MARKETO_URL, "rest/v1/", "leads.json?access_token={token}"]
+)
+
+
+class MarketoApi:
+    # Marketo isn't fast, so give it plenty of time to make a connection,
+    # and respond
+    api_session = api.requests.Session(timeout=(2, 12))
+    token = ""
+
+    def __init__(self):
+        self.authenticate()
+
+    def process_response(self, response):
+        try:
+            body = response.json()
+        except ValueError as decode_error:
+            api_error_exception = ApiResponseDecodeError(
+                "JSON decoding failed: {}".format(decode_error)
+            )
+            raise api_error_exception
+
+        if not response.ok:
+            if body.get("success") is False:
+                api_error_exception = ApiResponseErrorList(
+                    "The api returned a list of errors",
+                    body["errors"][0]["code"],
+                    map(lambda error: error["message"], body["errors"]),
+                )
+
+                raise api_error_exception
+            else:
+                raise ApiResponseError("Unknown error from api", 500)
+
+        return body
+
+    def authenticate(self):
```

Comment:
tests are failing because you are making an API call when initializing the application.

Can't we get the token without making this request? Otherwise we will need to think of a way to mock this API call on the setup of every test.

Reply:
I think maybe won't work. I think we might need to initialize and do the handshake on every request that requires it. Maybe you can have a look at the Marketo API, but to me it seems like the token lifetime is really short.

Reply:
@jpmartinspt 	I'll let you investigate